### PR TITLE
Updated TurtleCoin-Utils Dependency

### PR DESCRIPTION
**Pull request by brandonlehmann**
A whole bunch of stuff in TurtleCoin-Utils is going async. This then requires that this bubbles up through downstream packages. Here's a helping hand.

### CODE REVIEW

File: lib/CryptoWrapper.ts [hunk 1]
```diff
@@ -20,11 +20,10 @@ export async function generateKeyDerivation(
     }
 
     try {
-        const key = await TurtleCoinCrypto.generateKeyDerivation(
+        return await TurtleCoinCrypto.generateKeyDerivation(
```
Comment (zpalmtree) [inline lib/CryptoWrapper.ts:23]:
```suggestion
        return TurtleCoinCrypto.generateKeyDerivation(
```

File: lib/CryptoWrapper.ts [hunk 2]
```diff
@@ -93,11 +92,9 @@ export async function underivePublicKey(
     }
 
     try {
-        const key = await TurtleCoinCrypto.underivePublicKey(
+        return await TurtleCoinCrypto.underivePublicKey(
```
Comment (zpalmtree) [inline lib/CryptoWrapper.ts:95]:
```suggestion
        return TurtleCoinCrypto.underivePublicKey(
```

File: lib/Logger.ts [hunk 3]
```diff
@@ -15,11 +15,15 @@ class Logger {
 
     constructor(
         level?: LogLevel,
-        callback?: (prettyMessage: string, message: string, level: LogLevel, categories: LogCategory[]) => any) {
+        _callback?: (prettyMessage: string, message: string, level: LogLevel, categories: LogCategory[]) => any) {
```
Comment (zpalmtree) [inline lib/Logger.ts:18]: Why the `_`?

File: lib/Transfer.ts [hunk 4]
```diff
@@ -1226,8 +1234,8 @@ async function makeTransaction(
     let numGeneratedOnDemand = 0;
 
     const ourOutputs: Interfaces.Output[] = await Promise.all(ourInputs.map(async (input) => {
-        if (typeof input.input.privateEphemeral !== 'string' || !isHex64(input.input.privateEphemeral)) {
-            const [keyImage, tmpSecretKey] = await generateKeyImage(
+        if (input.input.privateEphemeral && !isHex64(input.input.privateEphemeral)) {
```
Comment (zpalmtree) [inline lib/Transfer.ts:1237]: This doesn't seem right to me. The private ephemeral is optional on the input, having potentially be generated before. If it's not present then we want to generate it now. Therefore I think we would want `if (!input.input.privateEphemeral)`

File: lib/WalletBackend.ts [hunk 5]
```diff
@@ -354,15 +365,17 @@ export class WalletBackend extends EventEmitter {
...
      * }
      * ```
+     * @param daemon
      * @param filename  The location of the wallet file on disk
      *
      * @param password  The password to use to decrypt the wallet. May be blank.
+     * @param config
      */
-    public static openWalletFromFile(
+    public static async openWalletFromFile(
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:374]: All functions which have example code above them (Generally all exported functions) will need updating to include an `await` where the return type is now a promise.

File: lib/WalletBackend.ts [hunk 6]
```diff
@@ -1506,20 +1538,7 @@ export class WalletBackend extends EventEmitter {
...
-            case 'wasmjs':
-                return true;
-            case 'unknown':
-                return true;
-            case 'js':
-                return false;
-            default:
-                return false;
-        }
+        return Crypto.isNative
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:1541]:
```suggestion
        return Crypto.isNative;
```

File: lib/WalletBackend.ts [hunk 7]
```diff
@@ -1477,7 +1509,7 @@ export class WalletBackend extends EventEmitter {
      * console.log(`Wallet is using the ${cryptoType} cryptographic library.`);
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:1509]: Example code/description is outdated

File: lib/Logger.ts (hunk 3 above)
Comment (brandonlehmann) [reply lib/Logger.ts:18]: Because it’s showing as an unused var in the method :)

File: lib/Transfer.ts (hunk 4 above)
Comment (brandonlehmann) [reply lib/Transfer.ts:1237]: You won’t always have the private spend key to do so... *wink*

File: lib/CryptoWrapper.ts (hunk 1 above)
Comment (brandonlehmann) [reply lib/CryptoWrapper.ts:23]: Missed it. Got it.

File: lib/CryptoWrapper.ts (hunk 2 above)
Comment (brandonlehmann) [reply lib/CryptoWrapper.ts:95]: Missed it. Got it.

File: lib/Logger.ts (hunk 3 above)
Comment (brandonlehmann) [reply lib/Logger.ts:18]: ah, rebase mess -- corrected

File: lib/WalletBackend.ts (hunk 6 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:1541]: got it

File: lib/WalletBackend.ts (hunk 5 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:374]: Updating...

File: lib/WalletBackend.ts (hunk 7 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:1509]: Updating...

File: lib/Transfer.ts (hunk 4 above)
Comment (brandonlehmann) [reply lib/Transfer.ts:1237]: Wait... I got you now, I update failure as the type is always `string` and we just wanna know if it exists or not.

File: lib/WalletBackend.ts [hunk 8]
```diff
@@ -1634,9 +1652,9 @@ export class WalletBackend extends EventEmitter {
 
             await this.daemon.init();
 
-            this.syncThread.start();
-            this.daemonUpdateThread.start();
-            this.lockedTransactionsCheckThread.start();
+            await this.syncThread.start();
+            await this.daemonUpdateThread.start();
+            await this.lockedTransactionsCheckThread.start();
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:1657]: We should either remove the `await`'s here or use a `Promise.all` so all 'threads' can start without having to wait for the previous one to finish initializing.

File: lib/WalletBackend.ts (hunk 7 above)
Comment (zpalmtree) [reply lib/WalletBackend.ts:1508]: Not seeing the update here?

File: lib/WalletBackend.ts [hunk 9]
```diff
@@ -1818,7 +1836,7 @@ export class WalletBackend extends EventEmitter {
      * }
      * ```
      */
-    public getMnemonicSeed(): ([string, undefined] | [undefined, WalletError]) {
+    public getMnemonicSeed(): Promise<([string, undefined] | [undefined, WalletError])> {
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:1839]: docstring needs updating to `await`

File: lib/WalletBackend.ts [hunk 10]
```diff
@@ -2829,7 +2846,7 @@ export class WalletBackend extends EventEmitter {
             }
 
             if (this.shouldPerformAutoOptimize && this.autoOptimize) {
-                this.performAutoOptimize();
+                await this.performAutoOptimize();
```
Comment (zpalmtree) [inline lib/WalletBackend.ts:2849]:
```suggestion
                this.performAutoOptimize();
```

This should remain non async. We don't want it to interrupt the periodic daemon updates.

File: lib/WalletBackend.ts (hunk 8 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:1657]: updated.

File: lib/WalletBackend.ts (hunk 9 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:1839]: handled

File: lib/WalletBackend.ts (hunk 10 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:2849]: Added new event `autoOptimizeError` that at least catches the error and emits it, how about that?

File: lib/WalletBackend.ts (hunk 7 above)
Comment (brandonlehmann) [reply lib/WalletBackend.ts:1521]: Hrrrmmm. Just saw this note but could have swore... checking...

Comment (brandonlehmann) [reply lib/WalletBackend.ts]: now it returns a string again

**Review by zpalmtree — APPROVED** [review]

**Pull request merged**